### PR TITLE
Master linkedin restapi msh

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
- Task: https://www.odoo.com/web?#id=8830&view_type=form&model=project.task&menu_id=3942&action=327
- Pad: https://pad.odoo.com/p/r.esmRR7w6RxXjBBlx

### CODE REVIEW

File: addons/web_linkedin/models/web_linkedin.py [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+        config_obj = self.env['ir.config_parameter']
+        apikey = self.api_key or ""
+        secret_key = self.secret_key or ""
+        config_obj.set_param("web.linkedin.apikey", apikey, groups=['base.group_users'])
+        config_obj.set_param("web.linkedin.secretkey", secret_key, groups=['base.group_users'])
+
+    @api.model
+    def create(self, vals):
+        if vals.get('api_key'):
+            self.env['linkedin'].check_valid_api_key(vals.get('api_key'))
```
Comment (mart-e) [inline addons/web_linkedin/models/web_linkedin.py]: Why not replace this by a constraint ?

File: addons/web_linkedin/models/web_linkedin.py [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+        self.env.user.sudo().write(data)
+
+    def need_authorization(self):
+        if not self.env.user.linkedin_token_validity or \
+                datetime.strptime(self.env.user.linkedin_token_validity.split('.')[0], DEFAULT_SERVER_DATETIME_FORMAT) < (datetime.now() + timedelta(minutes=1)):
+            return True
+        return False
+
+    def get_param_parameter(self, parameter):
+        return self.env['ir.config_parameter'].sudo().get_param(parameter)
```
Comment (mart-e) [inline addons/web_linkedin/models/web_linkedin.py]: Oh please don't do that. You just created a public method that allow to access any config parameter. :cry:

File: addons/web_linkedin/controllers/main.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    @http.route("/linkedin/linkedin_logout", type='json', auth='user')
+    def linkedin_logout(self, **post):
+        #Revoking access token from database, no need to invalidate token using API
+        #because there is no programmatic way to invalidate token(https://developer.linkedin.com/forum/revoke-authorization-access-token)
+        return request.env.user.sudo().write({'linkedin_token': False, 'linkedin_token_validity': False})
+
+    @http.route("/linkedin/sync_linkedin_contacts", type='json', auth='user')
+    def sync_linkedin_contacts(self, **post):
+        return request.env['linkedin'].with_context(post.get('local_context')).sync_linkedin_contacts(post.get('from_url'))
```
Comment (mart-e) [inline addons/web_linkedin/controllers/main.py]: maybe add some checks that these variables are present in `post`

File: addons/web_linkedin/models/web_linkedin.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+        companies = {}
+        universal_company = {}
+        warnings = []
+        if not params:
+            params = {'oauth2_access_token': self.env.user.linkedin_token}
+        if not headers:
+            headers = {'Content-type': 'application/json', 'Accept': 'text/plain', 'x-li-format': 'json'}
+        #search by universal-name
+        if kw.get('search_uid'):
+            universal_search_uri = "/companies/universal-name={company_name}:{company_fields}".format(company_name=kw['search_uid'], company_fields='(' + ','.join(COMPANY_FIELDS) + ')')
```
Comment (mart-e) [inline addons/web_linkedin/models/web_linkedin.py]: use urlparse to compose urls parameters (I don't think your code will work with companies special characters in the name)

File: addons/web_linkedin/models/web_linkedin.py [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+                'current-company': 'true',
+                'count': limit
+            }
+            people_criteria_uri = "/people-search:(people:{people_fields})".format(people_fields='(' + ','.join(PEOPLE_FIELDS) + ')')
+            status, res = self.send_request(people_criteria_uri, params=params, headers=headers, type="GET")
+            return res
+        else:
+            return {'status': 'need_auth', 'url': self._get_authorize_uri(from_url=from_url)}
+
+    def send_request(self, uri, pre_uri=BASE_API_URL, params={}, headers={}, type="GET"):
```
Comment (mart-e) [inline addons/web_linkedin/models/web_linkedin.py]: There is no existing method doing this ? If not, that should probably go in tools.

File: addons/web_linkedin/controllers/main.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import urllib2
+import werkzeug
+
+from openerp.addons.web import http
+from openerp.addons.web.http import request
+
+
+class Linkedin(http.Controller):
+
+    @http.route("/linkedin/authentication", type='http', auth="user")
```
Comment (rim-odoo) [inline addons/web_linkedin/controllers/main.py]: If this is a callback for linkedin server, are you convinced it will work with `auth="user"`?

**Pull request closed without merging**